zkmemory/src/supernova/memory_consistency_circuit.rs

```diff
@@ -0,0 +1,476 @@
+//! The inplementation uses the Supernova implementation of
+//! [here](https://github.com/argumentcomputer/arecibo/tree/dev/src)
+//! (They showed an example [here](https://github.com/argumentcomputer/arecibo/blob/dev/src/supernova/test.rs)
+//! but for other circuits, not memory consistency. We adapt for our use case.)
+//! The idea using Supernova is similar to our Nova's implementation
+//! [here](https://github.com/orochi-network/orochimaru/blob/main/zkmemory/src/nova/memory_consistency_circuit.rs)
+//! We let z_i to be the memory and each circuit Read or Write has
+//! a witness, which is the i-th trace (addr_i,val_i)
+//! The read circuit checks if z_i[add_i] == val_i, while the write
+//! circuit updates  z_i[add_i] := val_i and the commitment
+extern crate alloc;
+use alloc::vec::Vec;
+use arecibo::{
+    supernova::*,
+    traits::{CurveCycleEquipped, Dual, Engine},
+};
+use bellpepper_core::{num::AllocatedNum, ConstraintSystem, SynthesisError};
+use ff::PrimeField;
+use poseidon::poseidon_hash::ConstantLength;
+use poseidon::poseidon_hash::Hash;
+use poseidon::poseidon_hash::Spec;
+extern crate std;
+use core::marker::PhantomData;
+
+#[derive(Copy, Clone)]
+/// the trace record struct
+pub struct TraceRecord<F: PrimeField> {
+    address: F,
+    value: F,
+}
+
+#[derive(Clone)]
+/// The read circuit, used when the instruction equal to 0
+pub struct ReadCircuit<
+    F: PrimeField,
+    S: Spec<F, W, R> + Clone + core::marker::Sync + core::marker::Send,
+    const W: usize,
+    const R: usize,
+> {
+    trace: Option<TraceRecord<F>>,
+    next_instruction: Option<F>,
+    memory_size: Option<usize>,
+    _marker: PhantomData<S>,
+}
+
+impl<
+        F: PrimeField,
+        S: Spec<F, W, R> + Clone + core::marker::Sync + core::marker::Send,
+        const W: usize,
+        const R: usize,
+    > StepCircuit<F> for ReadCircuit<F, S, W, R>
+{
+    fn arity(&self) -> usize {
+        self.memory_size
+            .expect("failed to get arity of read circuit")
+            + 1
+    }
+    fn circuit_index(&self) -> usize {
```

Comment:
Must have sprace between method, and you missing inline document.

Reply:
I have fixed them